Consider gated pieces in SEE (I)

ceebo opened the pull request:
This is the version of my SEE patch that passed on the framework. It's quite ugly and took quite some debugging but the results look promising. I am also sending you an alternative version that is easier to understand, has the same bench signature but unfortunately requires the removal of `const` in quite a few random places. I doubt that the two versions differ in speed by very much so let me know your thoughts. Cheers.

Passed STC:

LLR: 2.95 (-2.94,2.94) [0.00,10.00]
Total: 1267 W: 451 L: 360 D: 456

and LTC:

LLR: 2.98 (-2.94,2.94) [0.00,10.00]
Total: 1074 W: 342 L: 258 D: 474

Bench: 1958670

ceebo commented:
(Closing and reopening to see if the Travis CI check will get re-run. I don't know why it failed first time)

ceebo commented:
(Failed again. I'm not sure what is happening with those Mac builds.)

ianfab commented:
I do not understand what is going wrong with mac builds either. I restarted the [build for current master](https://travis-ci.org/ianfab/Seirawan-Stockfish/builds/334179538) to check whether that would also fail now.

ianfab commented on src/position.cpp:
```diff
@@ -1156,12 +1157,33 @@ bool Position::see_ge(Move m, Value threshold) const {
+
+      // Create a copy of the piece bitboards array and put the gated
+      // piece in place of the original.
+      memcpy(tmpPiecesBB, byTypeBB, int(PIECE_TYPE_NB) * sizeof(Bitboard));
```
Don't you need to allocate the memory for `tmpPiecesBB` before doing the `memcpy`? (Sorry, I rarely use memcpy, so I am not expert).

ceebo replied:
Declaring `tmpPiecesBB[PIECE_TYPE_NB]` on line 1142 reserves space for `tmpPiecesBB` on the stack which will cease to be available when the function exits.

ianfab commented on src/position.cpp:
```diff
@@ -1156,12 +1157,33 @@ bool Position::see_ge(Move m, Value threshold) const {
 
   bool opponentToMove = true;
-  occupied = pieces() ^ from ^ to;
+  occupied = pieces() ^ to;
```
I would prefer to keep it as it is and check for the opposite condition below, since `NO_GATE_TYPE` is the default case.

ceebo committed:
Consider gated pieces in SEE

Passed STC:

LLR: 2.95 (-2.94,2.94) [0.00,10.00]
Total: 1267 W: 451 L: 360 D: 456

and LTC:

LLR: 2.98 (-2.94,2.94) [0.00,10.00]
Total: 1074 W: 342 L: 258 D: 474

Bench: 1958670

ceebo commented:
Those Mac builds failed again so I'm trying to swap the C-style memcpy for a C++-style std::copy...

EDIT: It doesn't seem to make any difference. I'm a bit stuck here. I can change back to memcpy if you prefer.

ceebo committed:
Define piecesBB as non-const

Need to apply a const_cast to byTypeBB to make this work.

Non-functional change

ceebo commented:
The last commit was a bit of a shot in the dark. It didn't help.

Also I noticed that it is only the 'debug' compiles that are failing on Mac, the optimised compiles further down the log seem fine. Sorry about this... this project is supposed to be a fun hobby!

ceebo committed:
Revert "Define piecesBB as non-const"

This reverts commit c67bf4e0948df04803718c0f3041974d382e3e5d.

ianfab commented:
No problem. That the linux builds succeed but the mac builds or more specifically only the non-optimized debug builds fail kind of suggests that there is undefined behavior, but I can not find anything suspicious in the code. But somehow both quite different versions of the PR failed in exactly the same way, so I guess something must be there.

Maybe it could have to do with special cases where SEE is called for strange moves. E.g., SEE can also be called for illegal moves, but only for pseudo-legal moves (unless I am missing something), so that should not be an issue. However, SEE can also be called for ["taking back" quiet moves in LMR](https://github.com/ianfab/Seirawan-Stockfish/blob/462c6ed3d4ac0387957bf06d73c5f2f33a006e52/src/search.cpp#L943). For gating moves, that might potentially be problematic, since the destination square is occupied by our own gated piece. But even if SEE returns nonsense results in these rare cases, I do not see how this could result in undefined behavior.

ceebo committed:
Modify signature.sh

ceebo commented:
Changed the signature script to output the full bench output on error. It's complaining about "Bus error 10" which I am in the process of Googling...

ceebo committed:
Copy byTypeBB by hand

ceebo committed:
Add asserts on a couple of array accesses

ceebo committed:
Abandon array swapping trick

ceebo committed:
Zero initialise tmpPiecesBB

ceebo commented:
Ugh... I just can't see where it is crashing. Not on an assertion failure so I can't think of how to track it down further...

ceebo commented:
Even stranger is that the bench appears to crash near the end of position 6 but position 6 doesn't have any pieces in hand in the start position!

ceebo committed:
Revert to original PR

Bench: 1958670

ceebo committed:
Set the tmpPieces array but don't use it

Bench: 2600408

ceebo committed:
Copy byTypeBB but don't amend it. Same b@nch

Bench: 2600408

ceebo committed:
Back to original but add some sanity asserts

Bench: 1958670

ceebo committed:
Pure original version but change zobrist keys

Bench: 1957123

ceebo commented:
The fact that 03c3564 passed and 8666fc0 failed is very interesting because the difference looks completely harmless.

The patch after that is also completely trivial and just involves changing the Zobrist random numbers... it looks like it will pass! Oh dear, the bug could be anywhere!

ianfab commented:
Thanks a lot for your analysis. So it looks like there is a bug unrelated to this PR that is only rarely triggered, and by chance your change happened to slightly change bench results in a way that the bug is triggered in one position. Sorry for the inconvenience, I will look into this as soon as possible. I would prefer to fix the issue before merging this PR if I am able to identify the cause in the next days.

ceebo commented:
Yes, no problem. I wonder what the easiest way to get a stack-trace is on Travis CI? I don't have any personal access to Mac's to test.

ceebo committed:
Crashing version but catch SIGBUS

ceebo committed:
Also catch SIGILL

Bench: 1958670

ceebo commented:
Tried to catch SIGBUS and print a stack trace but it didn't work particularly well. The error message changed to:

```
../tests/signature.sh: line 15:  1563 Illegal instruction: 4  ./stockfish bench >&bench.txt
```

Googling revealed plenty of stuff to do with OSX but hard to say if this is a problem with compile options or the code itself.

ianfab commented:
Debugging seems a bit difficult without access to a Mac for testing, so I ran a couple of deep searches (mainly on the bench positions) for different combinations of compilation options optimize=yes/no and debug=yes/no on linux in order to test whether I can maybe also find a reproducible crash on linux, but I did not observe any issues.

ceebo committed:
Try to run gdb in Travis on OSX

ceebo committed:
Change directory for core dumps

ceebo committed:
Bit of debugging re cores directory

ceebo committed:
List /cores dir near the end

ceebo committed:
cd /cores

ceebo committed:
Try lldb

ceebo commented:
Finally (!) got a stack trace with the last commit... possibly a crash in the endgame code?

```
(lldb) target create --core "./core.1563"
warning: (x86_64) /cores/./core.1563 load command 72 LC_SEGMENT_64 has a fileoff + filesize (0x2ea4a000) that extends beyond the end of the file (0x2ea49000), the segment will be truncated to match
warning: (x86_64) /cores/./core.1563 load command 73 LC_SEGMENT_64 has a fileoff (0x2ea4a000) that extends beyond the end of the file (0x2ea49000), ignoring this section
error: stockfish(0x0000000104450000) debug map object file '/Users/travis/build/ianfab/Seirawan-Stockfish/src/endgame.o' has changed (actual time is 0x5a71ba71, debug map time is 0x5a71ba12) since this executable was linked, file will be ignored
error: stockfish(0x0000000104450000) debug map object file '/Users/travis/build/ianfab/Seirawan-Stockfish/src/evaluate.o' has changed (actual time is 0x5a71ba72, debug map time is 0x5a71ba13) since this executable was linked, file will be ignored
Core file '/cores/./core.1563' (x86_64) was loaded.
(lldb) thread backtrace all
error: stockfish(0x0000000104450000) debug map object file '/Users/travis/build/ianfab/Seirawan-Stockfish/src/thread.o' has changed (actual time is 0x5a71ba77, debug map time is 0x5a71ba17) since this executable was linked, file will be ignored
error: stockfish(0x0000000104450000) debug map object file '/Users/travis/build/ianfab/Seirawan-Stockfish/src/uci.o' has changed (actual time is 0x5a71ba7a, debug map time is 0x5a71ba19) since this executable was linked, file will be ignored
error: stockfish(0x0000000104450000) debug map object file '/Users/travis/build/ianfab/Seirawan-Stockfish/src/main.o' has changed (actual time is 0x5a71ba72, debug map time is 0x5a71ba13) since this executable was linked, file will be ignored
error: stockfish(0x0000000104450000) debug map object file '/Users/travis/build/ianfab/Seirawan-Stockfish/src/search.o' has changed (actual time is 0x5a71ba78, debug map time is 0x5a71ba17) since this executable was linked, file will be ignored
* thread #1, stop reason = signal SIGSTOP
  * frame #0: 0x00007fffa4dfbbf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffa4ee77fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x00007fffa38694cd libc++.1.dylib`std::__1::condition_variable::wait(std::__1::unique_lock<std::__1::mutex>&) + 47
    frame #3: 0x00000001044cb731 stockfish`void std::__1::condition_variable::wait<Thread::wait_for_search_finished()::$_0>(std::__1::unique_lock<std::__1::mutex>&, Thread::wait_for_search_finished()::$_0) + 65
    frame #4: 0x00000001044cac6b stockfish`Thread::wait_for_search_finished() + 107
    frame #5: 0x00000001044d8fca stockfish`(anonymous namespace)::bench(Position&, std::__1::basic_istream<char, std::__1::char_traits<char> >&, std::__1::unique_ptr<std::__1::deque<StateInfo, std::__1::allocator<StateInfo> >, std::__1::default_delete<std::__1::deque<StateInfo, std::__1::allocator<StateInfo> > > >&) + 2762
    frame #6: 0x00000001044d50f6 stockfish`UCI::loop(int, char**) + 6614
    frame #7: 0x0000000104487e26 stockfish`main + 950
    frame #8: 0x00007fffa4ccd235 libdyld.dylib`start + 1
  thread #2, stop reason = signal SIGSTOP
    frame #0: 0x000000010446d480 stockfish`Endgame<(EndgameCode)2, Value>::operator()(Position const&) const + 32
    frame #1: 0x0000000104478e23 stockfish`Material::Entry::evaluate(Position const&) const + 51
    frame #2: 0x00000001044751bb stockfish`(anonymous namespace)::Evaluation<((anonymous namespace)::Trace::Tracing)0>::value() + 155
    frame #3: 0x00000001044750de stockfish`Eval::evaluate(Position const&) + 46
    frame #4: 0x00000001044c0c6e stockfish`Value (anonymous namespace)::qsearch<((anonymous namespace)::NodeType)0, false>(Position&, Search::Stack*, Value, Value, Depth) + 1422
    frame #5: 0x00000001044c038a stockfish`Value (anonymous namespace)::qsearch<((anonymous namespace)::NodeType)0, true>(Position&, Search::Stack*, Value, Value, Depth) + 2346
    frame #6: 0x00000001044c1390 stockfish`Value (anonymous namespace)::qsearch<((anonymous namespace)::NodeType)0, false>(Position&, Search::Stack*, Value, Value, Depth) + 3248
    frame #7: 0x00000001044c038a stockfish`Value (anonymous namespace)::qsearch<((anonymous namespace)::NodeType)0, true>(Position&, Search::Stack*, Value, Value, Depth) + 2346
    frame #8: 0x00000001044c1390 stockfish`Value (anonymous namespace)::qsearch<((anonymous namespace)::NodeType)0, false>(Position&, Search::Stack*, Value, Value, Depth) + 3248
    frame #9: 0x00000001044bcee2 stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 4562
    frame #10: 0x00000001044bf04a stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 13114
    frame #11: 0x00000001044bee43 stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 12595
    frame #12: 0x00000001044bf04a stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 13114
    frame #13: 0x00000001044bee43 stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 12595
    frame #14: 0x00000001044bf04a stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 13114
    frame #15: 0x00000001044bee43 stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 12595
    frame #16: 0x00000001044bf04a stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 13114
    frame #17: 0x00000001044bf04a stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 13114
    frame #18: 0x00000001044bf04a stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 13114
    frame #19: 0x00000001044bf04a stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 13114
    frame #20: 0x00000001044bf04a stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 13114
    frame #21: 0x00000001044bee43 stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 12595
    frame #22: 0x00000001044bf04a stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 13114
    frame #23: 0x00000001044bf04a stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 13114
    frame #24: 0x00000001044bf04a stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 13114
    frame #25: 0x00000001044bf04a stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 13114
    frame #26: 0x00000001044bf04a stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 13114
    frame #27: 0x00000001044bf04a stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 13114
    frame #28: 0x00000001044bf04a stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 13114
    frame #29: 0x00000001044bf04a stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 13114
    frame #30: 0x00000001044bf04a stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 13114
    frame #31: 0x00000001044bf04a stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 13114
    frame #32: 0x00000001044bf04a stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)0>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 13114
    frame #33: 0x00000001044b72ed stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)1>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 10573
    frame #34: 0x00000001044b76e5 stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)1>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 11589
    frame #35: 0x00000001044b76e5 stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)1>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 11589
    frame #36: 0x00000001044b76e5 stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)1>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 11589
    frame #37: 0x00000001044b76e5 stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)1>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 11589
    frame #38: 0x00000001044b76e5 stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)1>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 11589
    frame #39: 0x00000001044b76e5 stockfish`Value (anonymous namespace)::search<((anonymous namespace)::NodeType)1>(Position&, Search::Stack*, Value, Value, Depth, bool, bool) + 11589
    frame #40: 0x00000001044b0b56 stockfish`Thread::search() + 3574
    frame #41: 0x00000001044aea1e stockfish`MainThread::search() + 2478
    frame #42: 0x00000001044cab05 stockfish`Thread::idle_loop() + 277
    frame #43: 0x00000001044d0134 stockfish`void* std::__1::__thread_proxy<std::__1::tuple<void (Thread::*)(), Thread*> >(void*) + 532
    frame #44: 0x00007fffa4ee693b libsystem_pthread.dylib`_pthread_body + 180
    frame #45: 0x00007fffa4ee6887 libsystem_pthread.dylib`_pthread_start + 286
    frame #46: 0x00007fffa4ee608d libsystem_pthread.dylib`thread_start + 13
  thread #3, stop reason = signal SIGSTOP
    frame #0: 0x00007fffa4dfbbf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffa4ee77fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x00007fffa38694cd libc++.1.dylib`std::__1::condition_variable::wait(std::__1::unique_lock<std::__1::mutex>&) + 47
    frame #3: 0x00000001044cb781 stockfish`void std::__1::condition_variable::wait<Thread::idle_loop()::$_1>(std::__1::unique_lock<std::__1::mutex>&, Thread::idle_loop()::$_1) + 65
    frame #4: 0x00000001044caa83 stockfish`Thread::idle_loop() + 147
    frame #5: 0x00000001044d0134 stockfish`void* std::__1::__thread_proxy<std::__1::tuple<void (Thread::*)(), Thread*> >(void*) + 532
    frame #6: 0x00007fffa4ee693b libsystem_pthread.dylib`_pthread_body + 180
    frame #7: 0x00007fffa4ee6887 libsystem_pthread.dylib`_pthread_start + 286
    frame #8: 0x00007fffa4ee608d libsystem_pthread.dylib`thread_start + 13
(lldb) quit
```

ceebo committed:
Just run one failing test

ianfab commented:
@ceebo Thanks. So does the assertion
https://github.com/ianfab/Seirawan-Stockfish/blob/462c6ed3d4ac0387957bf06d73c5f2f33a006e52/src/endgame.cpp#L122
lead to the crash? That's strange.

ceebo commented:
Yes I was just about to write the same. Extremely weird. I don't see how it can fail there.

ceebo committed:
Print fen in bad position

ceebo committed:
More asserts

ceebo committed:
Print stuff in pos_is_ok

ceebo commented:
The output is not really changing. Just a crash on the first line of the KXK function no matter what...

ceebo committed:
More asserts

ceebo committed:
Try to print variables at crash

ceebo committed:
Remove assert from pos_is_ok

ceebo committed:
Print all fens after N KXK calls

ceebo committed:
Try to reduce stack usage

ceebo commented:
@ianfab It looks like the MacOS debug builds get a pitifully small amount of stack space. When running under Linux with commit bab55a0 the next printed FEN would be at depth 18 which is deeper in the tree than anything that went before.

It's not cool to put thumbs up to your own post but I don't care :)

ianfab commented:
@ceebo Thanks for debugging the issue. Unless you want to continue testing, please revert and squash to a version before the debugging, then I will merge.
